File: java/src/main/replicant/MappedResponder.java
// Copyright 2009 Kiel Hodges
package replicant;

import java.util.*;

import static replicant.collections.CollectionFunctions.*;

final class MappedResponder<ReturnValue> implements Responder<ReturnValue> {
  
  private final Map<Call, Result<ReturnValue>> responses = map();
  
  public void addResponseFor(Call call, Result<ReturnValue> response) {
    responses.put(call, response);
  }

  public Result<ReturnValue> resultFor(Call call) {
    if (responses.containsKey(call))
      return responses.get(call);
    
    StringBuffer message = 
      new StringBuffer("There is no response for ").append(call).append(".\nThere are responses for: ");
    for (Call expectedCall : responses.keySet())
      message.append("\n  ").append(expectedCall);
    throw new UnknownResponseException(message.toString());
  }
  
  public boolean equals(Object object) {
    if (object instanceof MappedResponder<?>) {
      MappedResponder<?> that = (MappedResponder<?>) object;
      return this.responses.equals(that.responses);
    }
    return false;
  }

  public int hashCode() {
    return 0;
  }
  
  
  public String toString() {
    return "MappedResponder(" + responses + ')';
  }

}

File: scala/src/main/scala/replicant/MappedResponder.scala
// Copyright 2011 Kiel Hodges
package replicant

class MappedResponder[Result] extends Responder[Result] {
  
  private val responses = scala.collection.mutable.LinkedHashMap[Call, () => Result]()
  private var called    = Set[Call]()

  def update(call: Call, response: () => Result) { responses(call) = response }
  
  def apply(call: Call): Response[Result] = {
    called += call
    responses.get(call) match {
      case Some(response) => ValueResponse(response) 
      case None           => UnknownResponse("No response expected for " + call)
    }
  }
  
  import org.scalatest.Assertions.assert

  def assertExpectationsMet { 
    val missingCalls = responses.keySet -- called
    assert(missingCalls.isEmpty, "Expected but did not receive " + Call.describe(missingCalls))
  }
  
  override def equals(other: Any) = other match {
    case that: MappedResponder[_] => true 
    case _ => false
  }

  override def hashCode = 37

}
  
